framework/src/service/ServiceReferenceBasePrivate.cpp

```diff
@@ -115,52 +126,42 @@ InterfaceMapConstPtr ServiceReferenceBasePrivate::GetServiceInterfaceMap(BundleP
 {
   InterfaceMapConstPtr s;
   if (!registration->available) return s;
-
   std::shared_ptr<ServiceFactory> serviceFactory;
-  int count = 0;
 
+  auto l = registration->Lock(); US_UNUSED(l);
```

Comment:
Agreed on removing the deadlock. I prematurely optimized for service creation.

I'm not sure if that will be sufficient to eliminate the race condition. I will take a closer look. 
In general, since ```ServiceReferenceBasePrivate::GetServiceInterfaceMap``` and ```ServiceReferenceBasePrivate::UngetService``` both modify the ```registration->dependents``` and ```registration->bundleServiceInstance``` data structures concurrently, all access (i.e. read and write) must be done within locks to guarantee no race conditions.